backend/cfg/cfg_ls.ml

```diff
@@ -0,0 +1,280 @@
+    | Some ({ begin_ = _; end_ = curr_end } as curr) ->
+      if (begin_ asr 1) - (curr_end asr 1) <= 1
+      then curr.end_ <- end_
+      else (
+        add_range reg curr;
+        Reg.Tbl.replace current_ranges reg { Range.begin_; end_ })
+  in
+  (* Equivalent to [update_interval_position_by_instr] in
+     "backend/interval.ml". *)
+  let update_instr :
+      type a.
+      int ->
+      a Cfg.instruction ->
+      trap_handler:bool ->
+      destroyed:Reg.t array ->
+      unit =
+   fun pos instr ~trap_handler ~destroyed ->
+    let on = pos lsl 1 in
+    let off = on + 1 in
+    if trap_handler
+    then
+      Array.iter Proc.destroyed_at_raise ~f:(fun reg ->
+          update_range reg ~begin_:on ~end_:on);
+    instr.ls_order <- on;
+    Array.iter instr.arg ~f:(fun reg -> update_range reg ~begin_:on ~end_:on);
+    Array.iter instr.res ~f:(fun reg -> update_range reg ~begin_:off ~end_:off);
+    let live = Cfg_dataflow.Instr.Tbl.find liveness instr.id in
+    Reg.Set.iter (fun reg -> update_range reg ~begin_:on ~end_:off) live.across;
+    Array.iter destroyed ~f:(fun reg -> update_range reg ~begin_:off ~end_:off)
+  in
+  let pos = ref 0 in
+  (* Equivalent to [walk_instruction] in "backend/interval.ml".*)
+  iter_instructions_order cfg_with_layout (Lazy.force Order.value)
+    ~instruction:(fun ~trap_handler instr ->
+      incr pos;
+      update_instr !pos instr ~trap_handler
+        ~destroyed:(Proc.destroyed_at_basic instr.desc))
+    ~terminator:(fun ~trap_handler term ->
+      incr pos;
+      update_instr !pos term ~trap_handler
+        ~destroyed:(Proc.destroyed_at_terminator term.desc);
+      incr pos);
+  Reg.Tbl.iter (fun reg (range : Range.t) -> add_range reg range) current_ranges;
+  Reg.Tbl.iter
+    (fun _reg (interval : Interval.t) ->
+      interval.ranges <- List.rev interval.ranges)
+    past_ranges;
+  if ls_debug && Lazy.force ls_verbose
+  then
+    iter_cfg_order cfg_with_layout (Lazy.force Order.value) ~f:(fun block ->
+        log ~indent:2 "(block %d)" block.start;
+        log_body_and_terminator ~indent:2 block.body block.terminator liveness);
+  State.update_intervals state past_ranges
+
+type spilling_reg =
+  | Spilling of Reg.t
+  | Not_spilling
+
+let allocate_stack_slot : Reg.t -> spilling_reg =
+ fun reg ->
+  log ~indent:3 "spilling register %a" Printmach.reg reg;
+  reg.spill <- true;
+  Spilling reg
+
+exception No_free_register
+
+let allocate_free_register : State.t -> Interval.t -> spilling_reg =
+ fun state interval ->
+  let reg = interval.reg in
+  match reg.loc, reg.spill with
+  | Unknown, true -> allocate_stack_slot reg
+  | Unknown, _ -> (
+    let reg_class = Proc.register_class reg in
+    let intervals = State.active state ~reg_class in
+    let first_available = Proc.first_available_register.(reg_class) in
+    match Proc.num_available_registers.(reg_class) with
+    | 0 -> fatal "register class %d has no available registers" reg_class
```

Comment:
It was intentional in the sense that it indeed felt
weird, as you say, to have an empty class. But we
can as well switch to `raise No_free_register`; this
is dead code and like you I think it will probably
stay that way.